network/noise/src/socket.rs

```diff
@@ -560,12 +603,22 @@ where
     ///
     /// Converts the noise session into transport mode and returns the NoiseSocket.
     fn finish(self) -> io::Result<NoiseSocket<TSocket>> {
-        let session = self
-            .0
-            .session
-            .into_transport_mode()
-            .map_err(|e| io::Error::new(io::ErrorKind::Other, format!("Noise error: {}", e)))?;
-        Ok(NoiseSocket { session, ..self.0 })
+        let session = self.0.session;
+        match session {
+            Session::Transport(_) => Err(io::Error::new(
+                io::ErrorKind::Other,
+                "Session not in handshake state".to_string(),
+            )),
+            Session::Handshake(sess) => {
+                let trans_sess = sess.into_transport_mode().map_err(|e| {
+                    io::Error::new(io::ErrorKind::Other, format!("Noise error: {}", e))
+                })?;
+                Ok(NoiseSocket {
+                    session: Session::Transport(Box::new(trans_sess)),
+                    ..self.0
+                })
```

Comment:
Maybe it would be a bit cleaner if we made this a method on our new Session Enum instead of having this logic here?